README.md
# imageinfo

extract image type and size from an inputstream

## Installation

add to your project dependencies
```
[imageinfo "0.0.2"]
```

## Usage

currently exposes only 1 function `info`:

```
(use 'imageinfo)

(info (io/input-stream "test/imageinfo/test_images/png/42x165.png"))

=> {:width 42, :height 165, :type :png}
```

## License

Copyright © 2016 Brian Hurlow

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-04-07
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-04-07
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/imageinfo/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/imageinfo/compare/0.1.0...0.1.1


test/imageinfo/core_test.clj
(ns imageinfo.core-test
  (:require [clojure.test :refer :all]
            [imageinfo.core :refer :all]
            [clojure.java.io :as io]))

;; ===== GIF

(defn files [path]
  (->> (java.io.File. path)
       (file-seq)
       (filter #(.isFile %))
       (filter #(not= ".DS_Store" (.getName %)))))

(defn file->dimensions [file]
  (-> (.getName file)
      (clojure.string/split #"\.")
      (first)
      (clojure.string/split #"x")
      (->> (map read-string))))

(deftest gif-dimensions
  (testing "getting the correct dimensions from a gif canvas"
    (doall
      (for [file (files "test/imageinfo/test_images/gif")]
        (let [expected-dimensions (file->dimensions file)
              ; msg (str "expected: " file expected-dimensions)
              stream (io/input-stream file)
              size (info stream)]
          (is (and (= (:width size) (first expected-dimensions)) 
                   (= (:height size) (second expected-dimensions)))))))))
                
;; ===== JEPG

(deftest jpeg-dimensions
  (testing "jpeg dimensions" 
    (doall
      (for [file (files "test/imageinfo/test_images/jpg")]
        (let [[exp-width exp-height] (file->dimensions file)
               stream (io/input-stream file)
               size (info stream)]
          (is (and (= (:width size) exp-width)
                   (= (:height size) exp-height))))))))

;; ===== PNG

(deftest png-dimensions
  (testing "png dimensions" 
    (doall
      (for [file (files "test/imageinfo/test_images/png")]
        (let [[exp-width exp-height] (file->dimensions file)
               stream (io/input-stream file)
               size (info stream)]
          (is (= (:width size) exp-width)))))))



test/imageinfo/network_test.clj
(ns imageinfo.network-test
  (:require [aleph.http :as http]
            [imageinfo.core :refer :all]))

;; I do this test manually

; (def url "https://avatars3.githubusercontent.com/u/1356376?v=3&s=460")

; (-> @(http/get url)
;      (:body)
;      (info))




src/imageinfo/core.clj
(ns imageinfo.core
  (:require [clojure.java.io :as io]))

;; ===== UTIL

;; TODO I forgot this was lazy...
(defn- read-bytes [stream n]
  (doall
    (for [x (range n)]
      (.read stream))))

(defn- square [x]
  (* x x))

(defn- u16-le
  "returns a little endian unsigned 16 bit integer"
  [a b]
  (bit-or
    (bit-shift-left b 8)
    a))

(defn- u16-be
  [byte-values]
  (.getInt
    (doto (java.nio.ByteBuffer/wrap (byte-array (concat [0 0] byte-values)))
      (.order java.nio.ByteOrder/BIG_ENDIAN))))

(defn- u32 [byte-values]
  (.getInt
    (doto (java.nio.ByteBuffer/wrap (byte-array byte-values))
      (.order java.nio.ByteOrder/BIG_ENDIAN))))

;; ===== PNG

(defn- read-png [first-six-bytes stream]
  ;; skip header info
  (read-bytes stream 10)
  {:width (u32 (read-bytes stream 4))
   :height (u32 (read-bytes stream 4))
   :type :png})

;; ===== JPEG

(def jpeg-sof
  {0xc0 true
   0xc1 true
   0xc2 true
   0xc3 true
   0xc5 true
   0xc6 true
   0xc7 true
   0xc9 true
   0xca true
   0xcb true
   0xcd true
   0xce true
   0xcf true}) 

(defn- read-jpg [first-six-bytes stream]
  (let [magick (read-bytes stream 2)]
    (loop [x stream]
      (let [byte (.read x)]
        (if (= -1 byte)
          false ;; done reading
          (if (= 0xff byte)
            (do 
              ;; if we hit marker, skip marker
              (read-bytes x 1)
              (recur x))
            (do
              ;; if non-sof jump to next marker
              (if (nil? (get jpeg-sof byte))
                (do 
                  (let []
                      ;; this advance is required
                      (read-bytes x 2)
                      ;; TODO needs to skip to the next marker here
                      ;; otherwise there's the possibility of reading an 
                      ;; SOF marker that is actually image data
                      (recur x)))
                (do 
                    ;; advance into to data frame
                    (read-bytes x 3)
                    {:height (u16-be (read-bytes x 2))
                     :width (u16-be (read-bytes x 2)) 
                     :type :jpg})))))))))

;; ===== GIFS 

;; close stream??
(defn- read-gif [first-six-bytes stream]
  (let [header first-six-bytes
        screen-descriptor (read-bytes stream 7)]
    {:width (u16-le (first screen-descriptor) (second screen-descriptor))
     :height (u16-le (nth screen-descriptor 2) (nth screen-descriptor 3))
     :type :gif}))

;; ===== INFO

(def jpg-signature
  '(0xff 0xd8))

(def gif-signature 
  '(0x47 0x49 0x46))

(def png-signature 
  '(0x89 0x50 0x4e 0x47 0x0D 0x0A))

;; This impl works and passes all tests but kinda sucks
;; I really want/need an immutable stream representation that doesn't
;; deplete on .read calls. Hence the first-six-bytes bullshit
(defn info [stream]
  (let [first-six-bytes (doall (read-bytes stream 6))]
    (cond
      (= jpg-signature (take 2 first-six-bytes)) (read-jpg first-six-bytes stream)
      (= gif-signature (take 3 first-six-bytes)) (read-gif first-six-bytes stream)
      (= png-signature first-six-bytes) (read-png first-six-bytes stream)
      :else nil)))



ImageInfo.java
/*
 * Taken from kickjava.com/src/imageinfo/ImageInfo.java.htm
 *
 * Original copyright:
 *
 * A Java class to determine image width, height and color depth for
 * a number of image file formats.
 *
 * Written by Marco Schmidt 
 * <http://www.geocities.com/marcoschmidt.geo/contact.html>.
 *
 * Contributed to the Public Domain.
 *
 * Last modification 2003-07-28
 * Read more: http://kickjava.com/src/imageinfo/ImageInfo.java.htm
 */
package imageinfo;

import java.io.FileInputStream;
import java.io.InputStream;
import java.io.IOException;
import java.net.URL;


/**
 *
 * @author achoy
 */
public class ImageInfo {

    public static final int FORMAT_GIF = 1;
    
    private int width;
    private int height;
    private int bitsPerPixel;
    private int format;
    private InputStream  in;
    private boolean collectComments = false;
    private boolean determineNumberOfImages;
    private int numberOfImages;
    private int tryCheckImages;

    public ImageInfo()
    {
        tryCheckImages = 5;
    }
    
    public boolean check() {
        format = -1;
        width = -1;
        height = -1;
        bitsPerPixel = -1;
         
        try
        {
            int b1 = read() & 0xff;
            int b2 = read() & 0xff;
            if (b1 == 0x47 && b2 == 0x49) {
                return checkGif();
            }
            return false;
        }
        catch (IOException ioe)
        {
            return false;
        }
    }
    
    private boolean checkGif() throws IOException 
    {
        final byte[] GIF_MAGIC_87A = {0x46, 0x38, 0x37, 0x61};
        final byte[] GIF_MAGIC_89A = {0x46, 0x38, 0x39, 0x61};
        byte[] a = new byte[11]; // 4 from the GIF signature + 7 from the global header
        if (read(a) != 11) {
            return false;
        }
        if ((!equals(a, 0, GIF_MAGIC_89A, 0, 4))
                && (!equals(a, 0, GIF_MAGIC_87A, 0, 4))) {
            return false;
        }
        format = FORMAT_GIF;
        width = getShortLittleEndian(a, 4);
        height = getShortLittleEndian(a, 6);
        int flags = a[8] & 0xff;
        bitsPerPixel = ((flags >> 4) & 0x07) + 1;
        if (!determineNumberOfImages) {
            return true;
        }
        // skip global color palette
        if ((flags & 0x80) != 0) {
            int tableSize = (1 << ((flags & 7) + 1)) * 3;
            skip(tableSize);
        }
        numberOfImages = 0;
        int blockType;
        int lWidth, lHeight;
        do {
            blockType = read();
            switch (blockType) {
                case (0x2c): // image separator
                {
                        
                    if (read(a, 0, 9) != 9) {
                        return false;
                    }
                    lWidth = getShortLittleEndian(a, 4);
                    lHeight = getShortLittleEndian(a, 6);                  
                    System.out.println("-----");
                    System.out.println("LWIDTH" + lWidth);
                    System.out.println("LHEITGHT" + lHeight);
                    if (lWidth != width) width = lWidth;
                    if (lHeight != height) height = lHeight;    
                    flags = a[8] & 0xff;
                    int localBitsPerPixel = (flags & 0x07) + 1;
                    if (localBitsPerPixel > bitsPerPixel) {
                        bitsPerPixel = localBitsPerPixel;
                    }
                    if ((flags & 0x80) != 0) {
                        skip((1 << localBitsPerPixel) * 3);
                    }
                    skip(1); // initial code length
                    int n;
                    do {
                        n = read();
                        if (n > 0) {
                            skip(n);
                        } else if (n == -1) {
                            return false;
                        }
                    } while (n > 0);
                    numberOfImages++;
                    
                    // try up to tryCheckImages images
                    if (tryCheckImages > 0 && numberOfImages >= tryCheckImages)
                        return true;

                    break;
                }
                case (0x21): // extension
                {
                    int extensionType = read();
                    if (collectComments && extensionType == 0xfe) {
                        StringBuffer sb = new StringBuffer();
                        int n;
                        do {
                            n = read();
                            if (n == -1) {
                                return false;
                            }
                            if (n > 0) {
                                for (int i = 0; i < n; i++) {
                                    int ch = read();
                                    if (ch == -1) {
                                        return false;
                                    }
                                    sb.append((char) ch);
                                }
                            }
                        } while (n > 0);
                    } else {
                        int n;
                        do {
                            n = read();
                            if (n > 0) {
                                skip(n);
                            } else if (n == -1) {
                                return false;
                            }
                        } while (n > 0);
                    }
                    break;
                }
                case (0x3b): // end of file
                {
                    break;
                }
                default: {
                    return false;
                }
            }            
        } while (blockType != 0x3b);
        return true;
    }
    
    private boolean equals(byte[] a1, int offs1, byte[] a2, int offs2, int num) {
        while (num-- > 0) {
            if (a1[offs1++] != a2[offs2++]) {
                return false;
            }
        }
        return true;
    }

    private int getShortBigEndian(byte[] a, int offs) {
         return
             (a[offs] & 0xff) << 8 | 
             (a[offs + 1] & 0xff);
    }

    private int getShortLittleEndian(byte[] a, int offs) {
        return (a[offs] & 0xff) | (a[offs + 1] & 0xff) << 8;     
    }

    private int read() throws IOException
    {
        if (in != null) return in.read();
        return -1;
    }
    
    private int read(byte[] a) throws IOException
    {
        if (in != null) return in.read(a);
        return -1;
    }
    
    private int read(byte[] a, int offset, int num) throws IOException
    {
        if (in != null) return in.read(a, offset, num);
        return -1;
    }
    
    private void skip(int num) throws IOException {
        while (num > 0) {
            long result = 0;
            if (in != null) {
                result = in.skip(num);
            } else {
                //result = din.skipBytes(num);
            }
            if (result > 0) {
                num -= result;
            }
        }
    }
    
    public void setInput(InputStream inStream) {
        in = inStream;
    }
    
    public void printVerbose(String sourceName)
    {
        System.out.println(
                "SourceName=" + sourceName +
                "; Width=" + width + 
                "; Height=" + height + 
                "; BitsPerPixel=" + bitsPerPixel
        );
    }
    
    public void processImage(String sourceName, InputStream in)
    {
        determineNumberOfImages = true;
        setInput(in);
        if (!check())
        {
            System.out.println("Invalid GIF image");
            return;
        }
        printVerbose(sourceName);
    }
    
    /**
     * @param args the command line arguments
     * 
     * Example command line:  x.gif https://website/image.gif
     */

    public static void main(String[] args) {
        
        args[0] = "tiny.gif";
        // args[0] = "keyboard.gif";
        ImageInfo imageInfo = new ImageInfo();
        if (args.length > 0)
        {
            int index = 0;
            while (index < args.length)
            {
                String name = args[index++];
                InputStream in = null;
                System.out.println("STARTING WITH YOOOO: >>" + name);
                try
                {
                    if (name.startsWith("http://") || name.startsWith("https://"))
                        in = new URL(name).openConnection().getInputStream();
                    else
                        in = new FileInputStream(name);
                    
                    imageInfo.processImage(name, in);
                }
               catch (Exception e)
                {
                    System.out.println("Error reading file: " + e.getMessage());
                }
                finally
                {
                    try
                    {
                        if (in != null)
                            in.close();
                    }
                    catch (Exception e) {}
                }
            }
        }
    }
    
}
